pkg/kubecost/allocation.go

```diff
@@ -71,6 +71,38 @@ type Allocation struct {
 	RAMCost                float64    `json:"ramCost"`
 	SharedCost             float64    `json:"sharedCost"`
 	ExternalCost           float64    `json:"externalCost"`
+
+	// RawAllocationOnly is a pointer so if it is not present it will be
+	// marshalled as null rather than as an object with Go default values.
+	RawAllocationOnly *RawAllocationOnlyData `json:"rawAllocationOnly"`
+}
+
+// RawAllocationOnlyData is information that only belong in "raw" Allocations,
+// those which have not undergone aggregation, accumulation, or any other form
+// of combination to produce a new Allocation from other Allocations.
+//
+// Max usage data belongs here because computing the overall maximum from two
+// or more Allocations is a non-trivial operation that cannot be defined without
+// maintaining a large amount of state. Consider the following example:
+// _______________________________________________
+//
+// A1 Using 3 CPU    ----      -----     ------
+// A2 Using 2 CPU      ----      -----      ----
+// A3 Using 1 CPU         ---       --
+// _______________________________________________
+//                   Time ---->
+//
+// The logical maximum CPU usage is 5, but this cannot be calculated iteratively,
+// which is how we calculate aggregations and accumulations of Allocations currently.
+// This becomes a problem I could call "maximum sum of overlapping intervals" and is
+// essentially a variant of an interval scheduling algorithm.
+//
+// If we had types to differentiate between regular Allocations and AggregatedAllocations
+// then this type would be unnecessary and its fields would go into the regular Allocation
+// and not in the AggregatedAllocation.
+type RawAllocationOnlyData struct {
```

Comment:
My only challenge here is: why not put these on the main `Allocation` type with a comment about zeroing them out on `AggregateBy` and a `json:"cpuCoreUsageMax,omitempty"` so that they aren't encoded they're zero?

That has the advantage of keeping the type flatter and simpler to interact with. It has the disadvantage of perhaps misleading one into thinking max usage is actually zero, when in fact it's just not set.

Reply:
I like the idea of keeping `Allocation` flatter like you've suggested, but it does overload 0: I'm not sure if it would ever happen, but I can imagine a legitimate `Allocation` that has actual 0 usage. When this is expanded to include variance, 0 is an even more legitimate and possible value.

I dislike the "pointer to a struct because of how Go does JSON marshaling" thing that I'm doing right now. I think the right alternative if we want to keep the type flatter is to stick a `bool` flag in it and check that instead of relying on Go's default values.

What do you think?

Reply:
The point about variance using 0 as a distinct and legitimate value is compelling. Argh. The bool flag is kinda nasty, too, from an internal-API perspective. (Imagine accessing one of those numbers, but not realizing that you were supposed to check a separate bool value first, just to be sure.)

AWS will often use types like `*string` or `*int` to give that extra `nil` option. We could do that, but... I also don't love that.

The only other idea that comes to mind is to make the fields flat and unexported, then provide accessor functions like `CPUCoreMaxUsage() (float64, error)` so that an error can be returned when accessing on an aggregated `Allocation`. Maybe better? Maybe over-engineered?

Reply:
Yeah I'm not a huge fan of just raw `*string`. Private fields and accessor functions is a decent idea, but it makes deserialization even more of a pain which currently just works https://github.com/kubecost/cost-model/blob/develop/pkg/kubecost/allocation_test.go#L434 I think without us writing a manual `UnmarshalJSON`.

I really don't like any of our options here. Thoughts?

Reply:
Well, it'd be a `*float64` in this case, but I take it that you still don't like that. `nil`s are a bummer, but I guess they're also kind of what we need (something to differentiate from `0`).

While I've never let free serialization stand between me and the design I want, it's also probably not the perfect solution. If it were me, I'd probably punt and just go with something simple (probably `*float64`) or perhaps I'd go with the accessor idea and punt on unmarshaling, but at this point I would just trust you to make the decision whatever that may be.

Reply:
I'm going to stick with what I've got right now because it tries to force (in a less-than-ideal way) the user to check that the allocation is unaggregated and the user only has to do so once and then can have confidence accessing the fields.